### ISSUE TITLE
Colspan problem on Datatable with hidden dynamic columns

### ISSUE THREAD

**Issue by Megachronix** [primefaces/primefaces#2726]
If you set `<p:columns visible="#{column.visible}">` dynamically in a DataTable, the calculated colspan of the empty message is calculated over every column, not just visible ones.

## 1) Environment
- PrimeFaces version: 6.1
- Application server + version: Tomcat 8.0.36 + Java 8
- Affected browsers: any browser

## 2) Expected behavior
If your filter does not match any value and the empty message is shown, the colspan is calculated only by visible columns.

## 3) Actual behavior
If your filter does not match any value and the empty message is shown, the colspan is calculated by any dynamic column.
![table](https://user-images.githubusercontent.com/31988044/30481749-8759788a-9a20-11e7-822c-2031d04a1b88.png)
![filter](https://user-images.githubusercontent.com/31988044/30481748-87555e6c-9a20-11e7-9ed0-3807e07f0202.png)

## 4) Steps to reproduce
Build & deploy example, filter id for 'z' (no car but the empty message should be displayed)

[primefaces-filter-dynamicColumns.zip](https://github.com/primefaces/primefaces/files/1306380/primefaces-filter-dynamicColumns.zip)

## 5) Sample XHTML
```html
<h:form id="form">
        <p:dataTable
            id="cars"
            var="car"
            value="#{dtColumnsView.cars}"
            widgetVar="carsTable"
            filteredValue="#{dtColumnsView.filteredCars}"
            emptyMessage="should be colspan=2">
            <p:columns
                visible="#{column.visible}"
                value="#{dtColumnsView.columns}"
                var="column"
                columnIndexVar="colIndex"
                sortBy="#{car[column.property]}"
                filterBy="#{car[column.property]}">
                <f:facet name="header">
                    <h:outputText value="#{column.header}" />
                </f:facet>
                <h:outputText value="#{car[column.property]}" />
            </p:columns>
        </p:dataTable>
    </h:form>
```

## 6) Sample bean
```java
@ManagedBean(name = "dtColumnsView")
@ViewScoped
public class ColumnsView implements Serializable {

    private String columnTemplate = "id brand year";

    private List<ColumnModel> columns;

    private List<Car> cars;

    private List<Car> filteredCars;

    @ManagedProperty("#{carService}")
    private CarService service;

    @PostConstruct
    public void init() {
        // "id", "brand", "year", "color", "price"
        cars = service.createCars(10);
        columns = new ArrayList<>();
        columns.add(new ColumnModel("Id", "id", true));
        columns.add(new ColumnModel("Brand", "brand", true));
        columns.add(new ColumnModel("Year", "year", false));
        columns.add(new ColumnModel("Color", "color", false));
        columns.add(new ColumnModel("Price", "price", false));

    }

    public List<Car> getCars() {
        return cars;
    }

    public List<Car> getFilteredCars() {
        return filteredCars;
    }

    public void setFilteredCars(List<Car> filteredCars) {
        this.filteredCars = filteredCars;
    }

    public void setService(CarService service) {
        this.service = service;
    }

    public String getColumnTemplate() {
        return columnTemplate;
    }

    public void setColumnTemplate(String columnTemplate) {
        this.columnTemplate = columnTemplate;
    }

    public List<ColumnModel> getColumns() {
        return columns;
    }

    static public class ColumnModel implements Serializable {

        private String header;
        private String property;
        private boolean visible;

        public ColumnModel(String header, String property, boolean visible) {
            this.header = header;
            this.property = property;
            this.visible = visible;
        }

        public String getHeader() {
            return header;
        }

        public String getProperty() {
            return property;
        }

        public boolean isVisible() {
            return visible;
        }

        public void setVisible(boolean visible) {
            this.visible = visible;
        }
    }
}
```

**Comment by stewe** [comment]
As a workaround you can add JavaScript right behind the DataTable and set the colspan manually. It's getting even easier, if you use a ColumnToggler:

```
<h:outputScript>
   $(function () {
      PF('columnTogglerWidgetVar').updateColspan();
   });
</h:outputScript>
```

Otherwise it will be something like
```
PF('carsTable').tbody.children('tr:first').children('td').attr('colspan', 2);
```
whereas you have to determine the correct number somehow. Maybe, take a look at: https://github.com/primefaces/primefaces/blob/25db20899902593b2858958ec25823414acee2a3/src/main/resources/META-INF/resources/primefaces/columntoggler/columntoggler.js#L422

**Comment by melloware** [comment]
This was fixed in 7.0.10 and 8.0 with this fix: https://github.com/primefaces/primefaces/issues/5211

Particularly calculateColSpan ignoring hidden columns
```javascript
calculateColspan: function() {
        var visibleHeaderColumns = this.thead.find('> tr:first th:not(.ui-helper-hidden)'),
            colSpanValue = 0;

        for(var i = 0; i < visibleHeaderColumns.length; i++) {
```
